ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PlanNode.java

```diff
@@ -192,10 +195,13 @@ static void throwKeysNotIncludedError(
       final LogicalSchema schema
   ) {
     // When doing a `select *` key columns should be at the front of the column list
-    // but are added at the back during processing for performance reasons.
-    // Switch them around here:
-    final Stream<Column> keys = columns.stream()
-        .filter(c -> schema.isKeyColumn(c.name()));
+    // but are added at the back during processing for performance reasons. Furthermore,
+    // they should be selected in the order of the key, if that is not the same as the
+    // ordering within the value. Switch them around here:
+    final ImmutableMap<ColumnName, Column> columnsByName = Maps.uniqueIndex(columns, Column::name);
+    final Stream<Column> keys = schema.key().stream()
```

Comment:
The main change (for the fix) seems to be to use `schema` instead of `columns` to find the keys?

Reply:
spot on, that's the first fix (the second fix is in the `SelectionUtil`